File: src/middleware/auth.middleware.js
const jwt = require('jsonwebtoken');
const db  = require('../config/database');

async function verifyToken(req, res, next) {
  const header = req.headers['authorization'];
  if (!header) return res.status(401).json({ message: 'Token required' });

  const token = header.split(' ')[1];
  if (!token) return res.status(401).json({ message: 'Invalid format' });

  try {
    req.user = jwt.verify(token, process.env.JWT_SECRET);
  } catch {
    return res.status(401).json({ message: 'Invalid or expired token' });
  }

  // RF-04: verify token was not invalidated by logout
  const { rows } = await db.query(
    'SELECT id FROM token_blacklist WHERE token = $1',
    [token]
  );
  if (rows.length > 0) {
    return res.status(401).json({ message: 'Session expired. Please log in again' });
  }

  next();
}

module.exports = { verifyToken };

File: src/middleware/auth_middleware.py
import os
from fastapi import HTTPException, Header
from jose import jwt, JWTError
from src.config.database import db

def verify_token(authorization: str = Header(...)):
    if not authorization or not authorization.startswith('Bearer '):
        raise HTTPException(status_code=401, detail='Token required')

    token = authorization.split(' ')[1]

    try:
        payload = jwt.decode(token, os.getenv('JWT_SECRET'), algorithms=['HS256'])
    except JWTError:
        raise HTTPException(status_code=401, detail='Invalid or expired token')

    # RF-04: verificar que el token no fue invalidado por logout
    row = db.token_blacklist.find_one({"token": token})
    if row:
        raise HTTPException(status_code=401, detail='Session expired. Please log in again')

    return payload
